Add admin page for HA tracker information.

cstyan opened the pull request:
It wasn't clear to me what the right way to serve the page was, and whether for other pages (ex: the ring pages) if we serve the page from every distributor/ingester or not.

Signed-off-by: Callum Styan <callumstyan@gmail.com>

cstyan commented:
cc @tomwilkie @gouthamve @jtlisi

tomwilkie commented on pkg/cortex/modules.go:
```diff
@@ -170,6 +170,7 @@ func (t *Cortex) initDistributor(cfg *Config) (err error) {
 
 	t.server.HTTP.HandleFunc("/all_user_stats", t.distributor.AllUserStatsHandler)
 	t.server.HTTP.Handle("/api/prom/push", t.httpAuthMiddleware.Wrap(http.HandlerFunc(t.distributor.PushHandler)))
+	t.server.HTTP.Handle("/ha-tracker", t.distributor.Replicas)
```
`Replicas` is nil is HA tracking is disabled, so this will panic.

tomwilkie replied:
In general, having this be nil when HA tracking is disabled is a pain.  I would always create the struct and internalise the enabled state within that struct.

tomwilkie commented on pkg/distributor/ha_tracker_http.go:
```diff
@@ -0,0 +1,81 @@
+		<h1>Cortex HA Tracker Status</h1>
+		<p>Current time: {{ .Now }}</p>
+		<form action="" method="POST">
+			<input type="hidden" name="csrf_token" value="$__CSRF_TOKEN_PLACEHOLDER__">
```
This is only needed if the page has an "action" - ie a button.  Same with the form.

tomwilkie commented on pkg/distributor/ha_tracker_http.go:
```diff
@@ -0,0 +1,81 @@
+						<th>User ID</th>
+						<th>Cluster</th>
+						<th>Replica</th>
+						<th>Timestamp</th>
```
When the lease runs out or when we last took the lease?  Both would be good.

tomwilkie commented on pkg/distributor/ha_tracker_http.go:
```diff
@@ -0,0 +1,81 @@
+						<td>{{ .UserID }}</td>
+						<td>{{ .Cluster }}</td>
+						<td>{{ .Replica }}</td>
+						<td>{{ .Timestamp }}</td>
```
Does this render in a useable form?

tomwilkie commented:
Looks really good - couple minor nits, but hopefully merge tomorrow.

tomwilkie commented on pkg/distributor/distributor.go:
```diff
@@ -166,13 +164,11 @@ func New(cfg Config, clientConfig ingester_client.Config, limits *validation.Ove
+	if err != nil {
+		return nil, err
 	}
+	d.Replicas = replicas
```
Nit: I prefer if this goes above the building of the Distributor now.

cstyan replied:
I'm assuming you mean move the creation of the HA tracker to above the creation of the distributor, but leave this assignment. I've done that in the latest commit.

tomwilkie commented on pkg/distributor/distributor.go:
```diff
@@ -109,7 +109,6 @@ type Config struct {
 	BillingConfig billing.Config             `yaml:"billing,omitempty"`
 	PoolConfig    ingester_client.PoolConfig `yaml:"pool,omitempty"`
 
-	EnableHATracker bool            `yaml:"enable_ha_tracker,omitempty"`
```
Not sure you can remove this completely - the distributor won't start if you don't provide consul/etc address for the ha tracker.

The idea was to move this flag into the HA tracker config instead, and have that struct 'know' if it were enabled or not - and not start the consul client, loop etc if it is.

cstyan replied:
:man_facepalming:

tomwilkie commented on pkg/distributor/ha_tracker_http.go:
```diff
@@ -0,0 +1,86 @@
+			Replica:      desc.Replica,
+			ElectedAt:    timestamp.Time(desc.ReceivedAt),
+			UpdateTime:   timestamp.Time(desc.ReceivedAt).Add(h.cfg.UpdateTimeout),
+			FailoverTime: timestamp.Time(desc.ReceivedAt).Add(h.cfg.FailoverTimeout),
```
Might be worth having these as durations, so they print out a bit more concisely?

Are UpdateTimeout and FailoverTimeout configurable per user?

cstyan replied:
Update/FailoverTimeout aren't configurable per user. Whether we accept HA samples from a user and what cluster/replica flags to look for are.

cstyan replied:
Durations via what call? `time.Until()`?

tomwilkie replied:
Yeah, `Until`.

tomwilkie commented:
Couple more nits.

cstyan commented:
@tomwilkie PTAL. In the most recent commit I've reverted the removal of/moved the `EnableHATracker` cfg flag to the HA tracker cfg from the Distributor cfg, added a prefix flag to the HA tracker flag registration (so we don't have to manually prefix each of it's flags with `distributor.`), and moved any checks for `if EnableHATracker` from distributor.go to ha_tracker.go. 

This means the HA Tracker struct is always created, but it only creates the client to it's KV store if `EnableHATracker == true`. Also if there's a misconfiguration of `EnableHATracker == false` but `acceptHASamples == true` for a user ID, we'll still accept samples that contain both HA tracking labels (cluster/replica), the HA tracker `checkReplica` will tell the distributor to accept any sample if `EnableHATracker == false`.

tomwilkie commented on pkg/distributor/ha_tracker.go:
```diff
@@ -80,27 +80,32 @@ type haTracker struct {
-// RegisterFlags adds the flags required to config this to the given FlagSet
-func (cfg *HATrackerConfig) RegisterFlags(f *flag.FlagSet) {
+// RegisterFlags adds the flags required to config this to the given FlagSet. If prefix is not an empty string it should end with a period.
+func (cfg *HATrackerConfig) RegisterFlags(prefix string, f *flag.FlagSet) {
```
TBH not sure the prefix is the right way to go here - its not like we're planning on using the HA tracker anywhere else, are we?

cstyan replied:
Not that I can think of, but it makes the flag registration a bit cleaner and less chance of new flags missing the `distributor.` prefix imo. I can revert this if you disagree.

tomwilkie replied:
We have to be really careful changing flag names.  For instance you dropped the dot between distributor and ha-tracker.  This has the potential to break rollouts and piss off users.

cstyan replied:
https://github.com/cortexproject/cortex/pull/1546/commits/8ffe4b9064485e5cc293892209bdb391c8257e9a#r312024987

cstyan commented:
test failures look like a flake?

bboreham commented:
`TestChunkStoreRandom` failures noted at #1522

It _could_ be a real problem, but not one introduced by this PR.

cstyan commented on pkg/distributor/distributor.go:
```diff
@@ -139,7 +139,7 @@ type Config struct {
 	cfg.BillingConfig.RegisterFlags(f)
 	cfg.PoolConfig.RegisterFlags(f)
-	cfg.HATrackerConfig.RegisterFlags(f)
+	cfg.HATrackerConfig.RegisterFlags("distributor.", f)
```
@tomwilkie I'm confused, I had the prefix as `distributor.` but your commit removing the prefix shows it was just `distributor`

tomwilkie replied:
`¯\_(ツ)_/¯`

gouthamve commented on pkg/distributor/ha_tracker_http.go:
```diff
@@ -0,0 +1,100 @@
+
+func (h *haTracker) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	h.electedLock.RLock()
+	defer h.electedLock.RUnlock()
```
Can we unlock right after we use it? I see we hold this lock until the network transfer is over and I see it as a potential DOS target, with someone deliberately holding the connection, as holding the lock will mean requests fail.
